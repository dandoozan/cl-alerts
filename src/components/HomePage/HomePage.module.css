.homePage {
}

.header {
  text-align: center;
  margin-bottom: 1.25em;
}
