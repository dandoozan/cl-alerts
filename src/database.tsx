import firebase from './firebase';

//tbx
function sleep(ms) {
  return new Promise((resolve, reject) => {
    setTimeout(resolve, ms);
  });
}

export async function addAlert(data) {
  console.log(`***in database addAlert, data=`, data);
  await sleep(1000);
  return '';
  // try {
  //   let db = firebase.firestore();
  //   let docRef = await db.collection('alerts').add(data);
  //   return docRef.id;
  // } catch (error) {
  //   console.error('Error adding document: ', error);
  // }
}
export async function getAlert(id) {
  console.log(`***in database getAlert, id=${id}`);
  await sleep(1000);
  try {
    let db = firebase.firestore();
    let doc = await db
      .collection('alerts')
      .doc(id)
      .get();
    if (doc.exists) {
      return doc.data();
    }
  } catch (error) {
    // console.error('Error getting document: ', error);
  }
  return null;
}
export async function updateAlert(id, data) {
  console.log(`***in database updateAlert, id=${id}, data=`, data);
  await sleep(1000);
  return false;
  // try {
  //    let db = firebase.firestore();
  //    await db
  //        .collection('alerts')
  //        .doc(id)
  //        .set(data);
  // } catch (error) {
  //    console.error('Error updating document: ', error);
  // }
}
export async function deleteAlert(id) {
  console.log(`***in database deleteAlert, id=${id}`);
  await sleep(1000);
  return true;
  // let db = firebase.firestore();
  // try {
  //   await db
  //     .collection('alerts')
  //     .doc(id)
  //     .delete();
  // } catch (error) {
  //   console.error('Error removing document: ', error);
  // }
}
