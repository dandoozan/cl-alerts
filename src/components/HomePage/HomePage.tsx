import React, { useState } from 'react';
import styles from './HomePage.module.css';
import { Link } from 'react-router-dom';
import { addAlert } from '../../database';
import HomePageForm from '../HomePageForm';
import SubmitError from '../SubmitError';
import { Modal, Button } from 'react-bootstrap';
import { defaultValues } from '../../formFields';

export default function HomePage(props) {
  let [alertData, setAlertData] = useState<any>({});
  let [submitError, setSubmitError] = useState(false);
  let [showModal, setShowModal] = useState(false);

  async function onSubmit(formValues, { setSubmitting }) {
    setSubmitError(false);
    let id = await addAlert(formValues);
    if (id) {
      setAlertData({ id, ...formValues });
      setShowModal(true);
    } else {
      setSubmitError(true);
    }
    setSubmitting(false);
  }

  function hideModal() {
    setShowModal(false);
  }

  return (
    <div className={styles.homePage}>
      <h2 className={styles.header}>Search Craigslist</h2>
      <HomePageForm initialValues={defaultValues} onSubmit={onSubmit} />
      {submitError && (
        <div className="text-sm-right">
          <SubmitError />
        </div>
      )}

      <Modal centered={true} show={showModal} onHide={hideModal}>
        <Modal.Header closeButton>
          <Modal.Title>Success!</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <div>
            <p>
              An alert has been created for <strong>{alertData.email}.</strong>
            </p>
            <p>
              You will automatically receive new craigslist results in your
              inbox according to the specified schedule.
            </p>
          </div>
        </Modal.Body>
        <Modal.Footer>
          <Link to={`/alert?id=${alertData.id}`}>
            <Button variant="secondary">View/Edit this alert</Button>
          </Link>
          <Button variant="primary" onClick={hideModal}>
            Done
          </Button>
        </Modal.Footer>
      </Modal>
    </div>
  );
}
