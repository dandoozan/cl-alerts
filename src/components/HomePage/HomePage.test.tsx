import React from 'react';
import HomePage from './HomePage';
import {
  render,
  fireEvent,
  wait,
  waitForElement,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import * as db from '../../database';
import { BrowserRouter } from 'react-router-dom';

//tests to make:
//  D-should call database.addAlert on submit
//  X-should disable button on submit
//  D-should show modal after submit
//  D-should show error message if addAlert failed
//  X-should re-enable submit button if addAlert failed
//  X-should hide error message when clicking submit after a failed submit
//  -should not submit if email is not valid
//  -should
//  -should
//  -should

beforeEach(() => {
  //@ts-ignore
  db.addAlert = jest.fn();
});

it('should call database.addAlert on submit', async () => {
  let { getByText } = render(
    <BrowserRouter>
      <HomePage />
    </BrowserRouter>
  );

  let submitButton = getByText('Submit');
  fireEvent.click(submitButton);

  //wait for the submit button to become enabled again
  await wait(() => expect(getByText('Submit')).not.toHaveAttribute('disabled'));

  expect(db.addAlert).toHaveBeenCalledTimes(1);
  //@ts-ignore (ignore because TS gives error: "Property 'mock' does not exist on type...")
  expect(db.addAlert.mock.calls[0]).toMatchSnapshot();
});
it('should show modal after submit', async () => {
  //@ts-ignore
  db.addAlert = jest.fn(() => 'ALERT_ID');

  let { getByText } = render(
    <BrowserRouter>
      <HomePage />
    </BrowserRouter>
  );

  let submitButton = getByText('Submit');
  fireEvent.click(submitButton);

  //wait for the success modal
  await waitForElement(() => getByText('Success!'));
});
it('should show error message if addAlert failed', async () => {
  let { getByText } = render(
    <BrowserRouter>
      <HomePage />
    </BrowserRouter>
  );

  let submitButton = getByText('Submit');
  fireEvent.click(submitButton);

  //wait for the submit button to become enabled again
  await wait(() => expect(getByText('Submit')).not.toHaveAttribute('disabled'));

  expect(getByText('Oops', { exact: false })).toBeInTheDocument();
});
