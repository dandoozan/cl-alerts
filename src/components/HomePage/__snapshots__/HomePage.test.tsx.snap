// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should call database.addAlert on submit 1`] = `
Array [
  Object {
    "category": "All categories",
    "city": "Phoenix",
    "days": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
    ],
    "email": "test@example.com",
    "hours": Array [
      3,
      7,
      11,
      15,
      19,
      23,
    ],
    "maxPrice": "",
    "minPrice": "",
    "picOnly": false,
    "searchTerm": "",
    "spamKeywords": "",
    "titleOnly": false,
  },
]
`;
